### PR TITLE
Migrate 'components/needs_team' module and associated tests to TS

### PR DESCRIPTION
**Pull request by kajumito**
## Summary
This PR migrates 'components/needs_team' module and associated tests to TypeScript

#### Ticket Link
Resolves [mattermost-server/issues/13728](https://github.com/mattermost/mattermost-server/issues/13728)

#### Comments
- I would suggest reviewer to watch should some of the props be optional.
- Didn't find @types for inbounce -library so I had to declare them myself

### CODE REVIEW

File: components/needs_team/needs_team.tsx [hunk 1]
```diff
@@ -1,64 +1,81 @@
...
-        history: PropTypes.object.isRequired,
-        teamsList: PropTypes.arrayOf(PropTypes.object),
+declare global {
+    interface Window {
+        isActive: boolean;
+    }
+}
+
+type Props = {
+    currentUser: {
```
Comment (devinbinnie) [inline components/needs_team/needs_team.tsx]: `currentUser` doesn't appear to be required in the original spec.

File: components/needs_team/needs_team.tsx [hunk 2]
```diff
@@ -1,64 +1,81 @@
...
-        super(params);
+type State = {
+    team: Team | null;
+    finishedFetchingChannels: boolean;
+    prevTeam: string;
+    teamsList: Team[];
+}
+
+export default class NeedsTeam extends React.Component<Props, State> {
+    public blurTime: number
```
Comment (devinbinnie) [inline components/needs_team/needs_team.tsx]: Nitpick, but semicolon at the end here?

File: components/needs_team/needs_team.tsx [hunk 3]
```diff
@@ -151,6 +168,10 @@ export default class NeedsTeam extends React.Component {
         window.removeEventListener('keydown', this.onShortcutKeyDown);
     }
 
+    shouldComponentUpdate() {
```
Comment (devinbinnie) [inline components/needs_team/needs_team.tsx]: Why was this added?

**Comment by lindy65** [comment]
Thanks @kajumito - I'll QA this once all dev reviews are done 👍 Please ping me here when ready! Thanks :)

File: components/needs_team/needs_team.tsx (hunk 3 above)
Comment (kajumito) [reply components/needs_team/needs_team.tsx]: There were following eslint error on this file:
Component is not optimized. Please add a shouldComponentUpdate method.eslint(react/require-optimization)

**Comment by kajumito** [comment]
> Looks pretty good overall, just a few comments.
> Also, can you make sure to resolve your merge conflict?
> Thanks!

**Comment by cpanato** [comment]
will check what is happening with the test server

File: components/needs_team/needs_team.tsx (hunk 3 above)
Comment (devinbinnie) [reply components/needs_team/needs_team.tsx]: If it's just a warning, we can safely ignore it.
Since this method isn't doing anything, we can remove it.

File: components/needs_team/needs_team.tsx (hunk 1 above)
Comment (kajumito) [reply components/needs_team/needs_team.tsx]: https://github.com/mattermost/mattermost-webapp/pull/4839/commits/6e4dd5c0e85debc5c48328b9b5a618bd6bc15a32

File: components/needs_team/needs_team.tsx (hunk 2 above)
Comment (kajumito) [reply components/needs_team/needs_team.tsx]: https://github.com/mattermost/mattermost-webapp/pull/4839/commits/6e4dd5c0e85debc5c48328b9b5a618bd6bc15a32

File: components/needs_team/needs_team.tsx (hunk 3 above)
Comment (kajumito) [reply components/needs_team/needs_team.tsx]: Done.

File: components/needs_team/needs_team.tsx [hunk 4]
```diff
@@ -83,17 +100,18 @@ export default class NeedsTeam extends React.Component {
...
             this.joinTeam(this.props);
         }
     }
 
-    static getDerivedStateFromProps(nextProps, state) {
+    static getDerivedStateFromProps(nextProps: Props, state: State) {
         if (state.prevTeam !== nextProps.match.params.team) {
-            const team = nextProps.teamsList ?
-                nextProps.teamsList.find((teamObj) =>
+            const team = state.teamsList ?
```
Comment (sudheerDev) [inline components/needs_team/needs_team.tsx]: Is this change intentional? Why are we changing this to state from props?

**Comment by lindy65** [comment]
Hi @sudheerDev - I'll QA once dev reviews are complete - please mention me here when ready for QA. Thanks!

**Comment by sudheerDev** [comment]
@lindy65 It is up for QA review.

**Comment by devinbinnie** [comment]
> Thanks @kajumito @sudheerDev - I couldn't determine from the changed files whether testing required is around users not belonging to a team? I've tested this area and didn't find any issues. Let me know if there is anything else that needs testing?

@lindy65 The `NeedsTeam` component is core to the main app view. So as long as the main app still works correctly, everything should be alright.
